tools/wasi/wasi.cpp

```diff
@@ -0,0 +1,217 @@
+    return fizzy::Value{uint32_t{UVWASI_ESUCCESS}};
+}
+
+bool run(int argc, const char** argv)
+{
+    // TODO: support both wasi_snapshot_preview1 and wasi_unstable?
+    const std::vector<fizzy::ImportedFunction> wasi_functions = {
+        {"wasi_snapshot_preview1", "proc_exit", {fizzy::ValType::i32}, std::nullopt,
+            wasi_proc_exit},
+        {"wasi_snapshot_preview1", "fd_read",
+            {fizzy::ValType::i32, fizzy::ValType::i32, fizzy::ValType::i32, fizzy::ValType::i32},
+            fizzy::ValType::i32, wasi_fd_read},
+        {"wasi_snapshot_preview1", "fd_write",
+            {fizzy::ValType::i32, fizzy::ValType::i32, fizzy::ValType::i32, fizzy::ValType::i32},
+            fizzy::ValType::i32, wasi_fd_write},
+        {"wasi_snapshot_preview1", "fd_prestat_get", {fizzy::ValType::i32, fizzy::ValType::i32},
+            fizzy::ValType::i32, wasi_fd_prestat_get},
+        {"wasi_snapshot_preview1", "fd_prestat_dir_name",
+            {fizzy::ValType::i32, fizzy::ValType::i32, fizzy::ValType::i32}, fizzy::ValType::i32,
+            wasi_return_enosys},
+        {"wasi_snapshot_preview1", "environ_sizes_get", {fizzy::ValType::i32, fizzy::ValType::i32},
+            fizzy::ValType::i32, wasi_environ_sizes_get},
+        {"wasi_snapshot_preview1", "environ_get", {fizzy::ValType::i32, fizzy::ValType::i32},
+            fizzy::ValType::i32, wasi_return_enosys},
+    };
+
+    uvwasi_options_t options = {
+        3,           // sizeof fd_table
+        0, nullptr,  // NOTE: no remappings
+        static_cast<uvwasi_size_t>(argc), argv,
+        nullptr,  // TODO: support env
+        0, 1, 2,
+        nullptr,  // NOTE: no special allocator
+    };
+
+    uvwasi_errno_t err = uvwasi_init(&state, &options);
+    if (err != UVWASI_ESUCCESS)
+    {
+        std::cerr << "Failed to initialise UVWASI: " << uvwasi_embedder_err_code_to_string(err)
+                  << "\n";
+        return false;
+    }
+
+    if (!std::filesystem::is_regular_file(argv[0]))
+    {
+        std::cerr << "File does not exists or is irregular: " << argv[0] << "\n";
+        return false;
+    }
+
+    std::ifstream wasm_file{argv[0]};
+    if (!wasm_file)
+    {
+        std::cerr << "Failed to open file: " << argv[0] << "\n";
+        return false;
+    }
+    const auto wasm_binary =
+        fizzy::bytes(std::istreambuf_iterator<char>{wasm_file}, std::istreambuf_iterator<char>{});
+
+    auto module = fizzy::parse(wasm_binary);
+    auto imports = fizzy::resolve_imported_functions(module, wasi_functions);
+    auto instance = fizzy::instantiate(
+        std::move(module), std::move(imports), {}, {}, {}, fizzy::MemoryPagesValidationLimit);
+    assert(instance != nullptr);
+
+    const auto start_function = fizzy::find_exported_function(instance->module, "_start");
+    if (!start_function.has_value())
+    {
+        std::cerr << "File is not WASI compatible (_start not found)\n";
```

Comment:
Not a job for now, but the wasi support should be designed as a library, otherwise it is difficult to test it.

Reply:
This case is not tested.